website/blog/2024/09-30-jsdom-based-scraping/index.md

```diff
@@ -0,0 +1,257 @@
+
+This article will discuss a new approach that we use in one of our Actors to obtain the authentication data from TikTok ads creative center generated by browser web applications without actually running the browser but instead of it, using JSDOM.
+
+![JSDOM based approach from scraping](./img/jsdom.webp)
+
+<!--truncate-->
+
+## Analyzing the website
+
+When you visit this URL:
+
+`https://ads.tiktok.com/business/creativecenter/inspiration/popular/hashtag/pc/en`
+
+You will see a list of hashtags with their live ranking, the number of posts they have, trend chart, creators, and analytics. You can also notice that we can filter the industry, set the time period, and use a check box to filter if the trend is new to the top 100 or not.
+
+![tiktok-trends](./img/tiktok-trends.webp)
+
+Our goal here is to extract the top 100 hashtags from the list with the given filters.
+
+The two possible approaches are to use [`CheerioCrawler`](https://crawlee.dev/docs/guides/cheerio-crawler-guide), and the second one will be browser-based scraping. Cheerio gives results faster but does not work with JavaScript-rendered websites.
+
+Cheerio is not the best option here as the [Creative Center](https://ads.tiktok.com/business/creativecenter/inspiration/popular/hashtag/pad/en) is a web application, and the data source is API, so we can only get the hashtags initially present in the HTML structure but not each of the 100 as we require.
+
+The second approach can be using libraries like Puppeteer, Playwright, etc, to do browser-based scraping and using automation to scrape all of the hashtags, but with previous experiences, it takes a lot of time for such a small task.
+
+Now comes the new approach that we developed to make this process a lot better than browser based and very close to CheerioCrawler based crawling.
+
+## JSDOM Approach
+
+:::note
+
+Before diving deep into this approach, I would like to give credit to [Alexey Udovydchenko](https://apify.com/alexey), Web Automation Engineer at Apify, for developing this approach. Kudos to him!
+
+:::
+
+In this approach, we are going to make API calls to `https://ads.tiktok.com/creative_radar_api/v1/popular_trend/hashtag/list` to get the required data.
+
+Before making calls to this API, we will need few required headers (auth data, so we will first make the call to `https://ads.tiktok.com/business/creativecenter/inspiration/popular/hashtag/pad/en`
+We will start this approach by creating a function that will create the URL for the API call for us and, make the call and get the data.
```

Comment:
you are missing a closing `)` here, and remove the line break (or add a blank line if this was supposed to be two paragraphs)